Pocket.Logger/Core/Formatter.cs

```diff
@@ -1,193 +1,312 @@
 
-                return new Formatter(t);
+        Formatter CreateFormatter(string t)
+        {
+            if (Interlocked.Increment(ref cacheCount) >= CacheLimit)
+            {
+                stopCaching = true;
             }
+
+            return new Formatter(t);
         }
+    }
 
-        internal class FormatterResult : IReadOnlyList<(string Name, object Value)>
+    internal class FormatterResult : IReadOnlyList<(string Name, object Value)>
+    {
+        private readonly StringBuilder formattedMessage;
+
+        public FormatterResult(StringBuilder formattedMessage)
         {
-            private readonly StringBuilder formattedMessage;
+            this.formattedMessage = formattedMessage;
+        }
 
-            public FormatterResult(StringBuilder formattedMessage)
-            {
-                this.formattedMessage = formattedMessage;
-            }
+        private readonly List<(string Name, object Value)> properties = new();
+
+        public void Add(string key, object value) => properties.Add((key, value));
+
+        public IEnumerator<(string Name, object Value)> GetEnumerator() => properties.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public int Count => properties.Count;
+
+        public (string Name, object Value) this[int index] => properties[index];
+
+        public override string ToString() => formattedMessage?.ToString() ?? "";
+    }
+}
+
+internal static partial class Format
+{
+    public static string ToLogString(
+       this in (
+           byte LogLevel,
+           DateTime TimestampUtc,
+           Func<(string Message, (string Name, object Value)[] Properties)> Evaluate,
+           Exception Exception,
+           string OperationName,
+           string Category,
+           (string Id,
+           bool IsStart,
+           bool IsEnd,
+           bool? IsSuccessful,
+           TimeSpan? Duration) Operation) e)
+    {
+        var (message, _) = e.Evaluate();
+
+        var logLevelString =
+            LogLevelString((LogLevel)e.LogLevel,
+                           e.Operation.IsStart,
+                           e.Operation.IsEnd,
+                           e.Operation.IsSuccessful,
+                           e.Operation.Duration);
+
+        return
+            $"{e.TimestampUtc:o} {e.Operation.Id.IfNotEmpty()}{e.Category.IfNotEmpty()}{e.OperationName.IfNotEmpty()} {logLevelString} {message} {e.Exception}";
+    }
+
+    static partial void CustomizeLogString(object value, ref string output);
+
+    internal static string ToLogString(this object value)
+    {
+        string output = null;
+
+        CustomizeLogString(value, ref output);
+
+        if (output is not null)
+        {
+            return output;
+        }
+
+        if (value is IEnumerable enumerable and not string)
+        {
+            return $"[ {string.Join(", ", enumerable.Cast<object>())} ]";
```

Comment:
Should the log string for elements of an enumerable also be customizable?

Reply:
This code isn't complex or interesting enough to start adding that much granularity, I think. Customizing via `CustomizeLogString` is enough.

Reply:
Ah as we discussed offline, what I meant is that the customization performed in `CustomizeLogString` is not being applied over here if the object appears inside an enumerable. It probably should

I am guessing there are more cases than just a simple enumerable too - such as dictionary values etc. :) We probably don't want to handle all such type traversal cases - but for the above case it seems reasonable to call `.ToLogString()` on the collection elements here...

Reply:
I think this immediately gets too complicated and the existing log output works fine. People can add that if they want to using this customization mechanism.

I do think I'd like to make this a little safer, though.